File: src/tvision4c/MessageEvent.cpp
#include "MessageEvent.h"

EXPORT tv::Error TV_MessageEvent_new(MessageEvent** out) {
    return tv::checkedNew(out);
}

EXPORT tv::Error TV_MessageEvent_delete(MessageEvent* self) {
    return tv::checkedDelete(self);
}

EXPORT tv::Error TV_MessageEvent_equals(MessageEvent* self, MessageEvent* other, BOOL* out) {
    if (!out) {
        return tv::Error_ArgumentNull;
    }

    if (!self && !other) {
        *out = TRUE;
        return tv::Success;
    }

    if (!self || !other) {
        *out = FALSE;
        return tv::Success;
    }

    if (self->command != other->command || self->infoPtr != other->infoPtr) {
        *out = FALSE;
        return tv::Success;
    }

    *out = TRUE;
    return tv::Success;
}

EXPORT tv::Error TV_MessageEvent_hash(MessageEvent* self, int32_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    int32_t result = 0;
    tv::hash(self->command, &result);
    tv::hash(self->infoPtr, &result);

    *out = result;
    return tv::Success;
}

EXPORT tv::Error TV_MessageEvent_get_command(MessageEvent* self, uint16_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->command;
    return tv::Success;
}

EXPORT tv::Error TV_MessageEvent_set_command(MessageEvent* self, uint16_t value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->command = value;
    return tv::Success;
}

EXPORT tv::Error TV_MessageEvent_get_infoPtr(MessageEvent* self, void** out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->infoPtr;
    return tv::Success;
}

EXPORT tv::Error TV_MessageEvent_set_infoPtr(MessageEvent* self, void* value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->infoPtr = value;
    return tv::Success;
}

File: src/TurboVision/System/MessageEvent.cs
using System.Runtime.InteropServices;

namespace TurboVision.System;

public unsafe partial class MessageEvent(void* ptr, bool owned, bool placement)
    : NativeObject<MessageEvent>(ptr, owned, placement)
{
    private sealed class Factory : NativeObjectFactory<Factory>
    {
        public Factory()
            : base(
                NativeMethods.TV_MessageEvent_placementSize,
                NativeMethods.TV_MessageEvent_placementNew,
                NativeMethods.TV_MessageEvent_new
            ) { }
    }

    public static int PlacementSize => Factory.Instance.PlacementSize;

    public MessageEvent(byte* placement)
        : this(Factory.Instance.PlacementNew(placement), owned: true, placement: true) { }

    public MessageEvent()
        : this(Factory.Instance.New(), owned: true, placement: false) { }

    protected override void PlacementDeleteCore(void* ptr)
    {
        TurboVisionException.Check(NativeMethods.TV_MessageEvent_placementDelete(ptr));
    }

    protected override void DeleteCore()
    {
        TurboVisionException.Check(NativeMethods.TV_MessageEvent_delete(Ptr));
    }

    protected override bool EqualsCore(MessageEvent other)
    {
        TurboVisionException.Check(
            NativeMethods.TV_MessageEvent_equals(Ptr, other.Ptr, out var equals)
        );
        return equals;
    }

    protected override int GetHashCodeCore()
    {
        TurboVisionException.Check(NativeMethods.TV_MessageEvent_hash(Ptr, out var hash));
        return hash;
    }

    public ushort Command
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_MessageEvent_get_command(Ptr, out var command)
            );
            return command;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_MessageEvent_set_command(Ptr, value));
        }
    }

    public void* InfoPtr
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(
                NativeMethods.TV_MessageEvent_get_infoPtr(Ptr, out var infoPtr)
            );
            return infoPtr;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_MessageEvent_set_infoPtr(Ptr, value));
        }
    }

    internal static partial class NativeMethods
    {
        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_placementSize(
            out int outSize,
            out int outAlignment
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_placementNew(byte* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_placementDelete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_new(out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_delete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_equals(
            void* self,
            void* other,
            [MarshalAs(UnmanagedType.I4)] out bool @out
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_hash(void* self, out int @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_get_command(void* self, out ushort @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_set_command(void* self, ushort value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_get_infoPtr(void* self, out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_MessageEvent_set_infoPtr(void* self, void* value);
    }
}
